homeassistant/components/media_player/songpal.py

```diff
@@ -0,0 +1,250 @@
+        endpoint = discovery_info["properties"]["endpoint"]
+        _LOGGER.debug("Got autodiscovered %s - endpoint: %s", name, endpoint)
+
+        device = SongpalDevice(name, endpoint)
+    else:
+        name = config.get(CONF_NAME)
+        endpoint = config.get(CONF_ENDPOINT)
+        device = SongpalDevice(name, endpoint)
+
+    try:
+        await device.initialize()
+    except SongpalException as ex:
+        _LOGGER.error("Unable to get methods from songpal: %s", ex)
+        raise PlatformNotReady
+
+    hass.data[PLATFORM][endpoint] = device
+
+    async_add_devices([device], True)
+
+    async def async_service_handler(service):
+        """Service handler."""
+        entity_id = service.data.get("entity_id", None)
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+
+        for device in hass.data[PLATFORM].values():
+            if device.entity_id == entity_id or entity_id is None:
+                _LOGGER.debug("Calling %s (entity: %s) with params %s",
+                              service, entity_id, params)
+
+                await device.async_set_sound_setting(params[PARAM_NAME],
+                                                     params[PARAM_VALUE])
+
+    hass.services.async_register(
+        DOMAIN, SET_SOUND_SETTING, async_service_handler,
+        schema=SET_SOUND_SCHEMA)
+
+
+class SongpalDevice(MediaPlayerDevice):
+    """Class representing a Songpal device."""
+
+    def __init__(self, name, endpoint):
+        """Init."""
+        import songpal
+        self._name = name
+        self.endpoint = endpoint
+        self.dev = songpal.Protocol(self.endpoint)
+        self._sysinfo = None
+
+        self._state = False
+        self._available = False
+        self._initialized = False
+
+        self._volume_control = None
+        self._volume_min = 0
+        self._volume_max = 1
+        self._volume = 0
+        self._is_muted = False
+
+        self._sources = []
+
+    async def initialize(self):
+        """Initialize the device."""
+        await self.dev.get_supported_methods()
+        self._sysinfo = await self.dev.get_system_info()
+
+    @property
+    def name(self):
+        """Return name of the device."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Return an unique ID."""
+        return self._sysinfo.macAddr
+
+    @property
+    def available(self):
+        """Return availability of the device."""
+        return self._available
+
+    async def async_set_sound_setting(self, name, value):
+        """Change a setting on the device."""
+        await self.dev.set_sound_settings(name, value)
+
+    async def async_update(self):
+        """Fetch updates from the device."""
+        from songpal import SongpalException
+        try:
+            volumes = await self.dev.get_volume_information()
+            if not volumes:
+                _LOGGER.error("Got no volume controls, bailing out")
```

Comment:
Should this set available to `False`?

Reply:
It should never get that far I suppose, but I can add a safe-guard for it.